sql: add disk backed indexed row container

yuzefovich opened the pull request:
Adds a wrapper around diskBackedRowContainer that stores an index
of the row as the last column of each row. These indices are needed
when computing window functions since each window function can have
its own ordering, and the indices are the way to "compose" the output
row.

Addresses: #32226.

Having the cache when we spill to disk makes huge difference:
```
BenchmarkDiskBackedIndexedRowContainer/InMemory-12             20000000           101 ns/op      78.71 MB/s
BenchmarkDiskBackedIndexedRowContainer/OnDiskWithCache-12       5000000           361 ns/op      22.12 MB/s
BenchmarkDiskBackedIndexedRowContainer/OnDiskWithoutCache-12               500000         39870 ns/op       0.20 MB/s
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/34196)
<!-- Reviewable:end -->

yuzefovich committed:
sql: add disk backed indexed row container

Adds a wrapper around DiskBackedRowContainer that stores an index
of the row as the last column of each row. These indices are needed
when computing window functions since each window function can have
its own ordering, and the indices are the way to "compose" the output
row.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/distsqlrun/row_container.go, line 615 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXKufwm-tYYVVQCe3gL:-LXKufwm-tYYVVQCe3gM:b-6g7gs8) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc23515cc8209af2c9ebb739b6516255dd32cb3d/pkg/sql/distsqlrun/row_container.go#L615)):*
> ```Go
> }
> 
> const maxIndexedRowsCacheSize = 4096
> ```

This number is somewhat arbitrary (I picked it based on the trackernetwork database and their query). I'm thinking maybe this should be a cluster setting? Any ideas/suggestions are welcome.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
I still need to review the actual meat of the algorithm, here's a first cut

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-:-LXZWK2M52evcUkS6VlR:b1vimp1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/rowcontainer/disk_row_container.go, line 183 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZTbC65N-r0z4G-V_D:-LXZTbC70wvcPdVJgi9o:b6drxmx) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/disk_row_container.go#L183)):*
> ```Go
> 	ctx context.Context, ordering sqlbase.ColumnOrdering,
> ) (SortableRowContainer, error) {
> 	// We need to create a new DiskRowContainer since it's ordering can only be
> ```

nit: s/it's/its/
___
*[pkg/sql/rowcontainer/disk_row_container.go, line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZTf2xE3X0Bm6hBtif:-LXZTf2xE3X0Bm6hBtig:b4lxsqf) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/disk_row_container.go#L203)):*
> ```Go
> 	}
> 	// TODO(yuzefovich): should it be called here?
> 	d.Close(ctx)
> ```

Hmm... it depends on whether or not you expect the caller to maintain references to the old container. It seems like the contract of this method is that you don't expect that - normally, the old container is made afresh and re-sorted.

Perhaps what you should do here is, instead of returning a brand new DiskRowContainer and exposing that complexity in the method's return values, close the old container and copy the new container's fields into the old container. So then the user could keep on using the old container pointer without being any wiser that some swapping happened out from under them. Does that make sense?

Essentially you'd say something like:

```
d.Close(ctx)
*d = newContainer
```

Then you could also remove the "new container" return value completely, i think.
___
*[pkg/sql/rowcontainer/row_container.go, line 529 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZUQxD2hDFUjL6T4F_:-LXZUQxD2hDFUjL6T4Fa:b-ylaidg) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L529)):*
> ```Go
> 	storedTypes   []sqlbase.ColumnType
> 	datumAlloc    sqlbase.DatumAlloc
> 	rowAlloc      sqlbase.EncDatumRowAlloc
> ```

Do you need these two fields? Can you use the equivalent ones from within DiskBackedRowContainer?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/rowcontainer/disk_row_container.go, line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZTf2xE3X0Bm6hBtif:-LX_0ngH8tZeqSEFXUYE:bjjivlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/disk_row_container.go#L203)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Hmm... it depends on whether or not you expect the caller to maintain references to the old container. It seems like the contract of this method is that you don't expect that - normally, the old container is made afresh and re-sorted.

Perhaps what you should do here is, instead of returning a brand new DiskRowContainer and exposing that complexity in the method's return values, close the old container and copy the new container's fields into the old container. So then the user could keep on using the old container pointer without being any wiser that some swapping happened out from under them. Does that make sense?

Essentially you'd say something like:

```
d.Close(ctx)
*d = newContainer
```

Then you could also remove the "new container" return value completely, i think.
</blockquote></details>

I like that idea, it would also make sense wrt to the function name.
___
*[pkg/sql/rowcontainer/row_container.go, line 140 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZbZgL4r8IgW5-E0M3:-LXZbZgL4r8IgW5-E0M4:b-j32uam) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L140)):*
> ```Go
> ) {
> 	acc := mon.MakeBoundAccount()
> 	mc.RowContainer.Init(acc, sqlbase.ColTypeInfoFromColTypes(types), rowCapacity)
> ```

I would make this `rowCapacity` change a different commit to keep changes contained but up to you
___
*[pkg/sql/rowcontainer/row_container.go, line 459 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_1R1--GKcZ-9C1Wo5:-LX_1R106tG7yvd5r54m:b-2d1fj4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L459)):*
> ```Go
> 		f.drc.Close(ctx)
> 	}
> 	if f.mrc != nil {
> ```

Why this change? `f.mrc` should never be nil, right?
___
*[pkg/sql/rowcontainer/row_container.go, line 521 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_3LUm3pGJuGjoNLPL:-LX_3LUm3pGJuGjoNLPM:bbvr4cx) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L521)):*
> ```Go
> // DiskBackedIndexedRowContainer is a wrapper around DiskBackedRowContainer
> // that adds an index to each row added in the order of addition of those rows
> // by storing an extra int column at the end of each row.
> ```

As stated, the `DiskRowContainer` does this. Maybe the comment needs to be expanded?
___
*[pkg/sql/rowcontainer/row_container.go, line 646 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZoV137UjCuqbH1yUO:-LXZoV137UjCuqbH1yUP:br5jnxl) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L646)):*
> ```Go
> // optimizations of maintaining a cache of tree.IndexedRow's and storing a disk
> // iterator along with the index of the row it currently points at.
> func (f *DiskBackedIndexedRowContainer) GetRow(pos int) tree.IndexedRow {
> ```

As I understand it the problem statement is this (focusing on disk): a row container must be created with an associated row ordinal and an ordering. The `DiskRowContainer` already does this with `rowID` and a supplied ordering. The difficulty however then comes from wanting to change the ordering on the `DiskRowContainer` but preserve the initial ordinal. I'm wondering if maybe we could rewrite keys in the DiskRowContainer, leaving the ordinal untouched (it is added at the end of the key). By rewrite I mean, when `ChangeOrdering` is called on the `DiskRowContainer`, we iterate over the `DiskRowContainer` and reinsert the row according to the given ordering. Don't know how possible it is.

Also, I think this could be a property observed on a normal `DiskBackedRowContainer`, maybe eliminating the need to create a new row container? Haven't thought much about how these guarantees would translate to memory row  containers.
___
*[pkg/sql/rowcontainer/row_container.go, line 679 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZmwVs2J7zcLJ0G6dI:-LXZmwVs2J7zcLJ0G6dJ:b-rcx8kq) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L679)):*
> ```Go
> 		for ; ; f.diskRowIter.Next() {
> 			if ok, err := f.diskRowIter.Valid(); err != nil {
> 				panic(err)
> ```

I would have this function return an error rather than panic
___
*[pkg/sql/rowcontainer/row_container.go, line 683 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_53pw7SbJao83Y7R5:-LX_53pw7SbJao83Y7R6:b-236wus) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L683)):*
> ```Go
> 				panic(fmt.Sprintf("row at pos %d not found", pos))
> 			}
> 			if f.idxRowIter == f.nextPosToCache {
> ```

How will this function be used? I wonder if it makes sense to just call `Next()` `n` times. We also have to be careful about the cache's memory overhead. 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/rowcontainer/disk_row_container.go, line 183 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZTbC65N-r0z4G-V_D:-LX_ALjK9C1qS3SskNkU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/disk_row_container.go#L183)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

nit: s/it's/its/
</blockquote></details>

Done.
___
*[pkg/sql/rowcontainer/disk_row_container.go, line 203 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZTf2xE3X0Bm6hBtif:-LX_DtrXDKqXZpSOXm1w:bt690a2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/disk_row_container.go#L203)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I like that idea, it would also make sense wrt to the function name.
</blockquote></details>

Great, done. I remember trying something similar when I wrote the code, but the fields were not copied (I probably had a value receiver so the container wasn't updated properly).
___
*[pkg/sql/rowcontainer/row_container.go, line 140 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZbZgL4r8IgW5-E0M3:-LX_Ju5k74iWjA4IVB2S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L140)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would make this `rowCapacity` change a different commit to keep changes contained but up to you
</blockquote></details>

Done.
___
*[pkg/sql/rowcontainer/row_container.go, line 459 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_1R1--GKcZ-9C1Wo5:-LX_K3_29_7LfZY-e1JJ:bbea45q) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L459)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why this change? `f.mrc` should never be nil, right?
</blockquote></details>

Nice catch, thanks. At some point, I added a flag `shouldUseDiskFromStart` which would tell the `diskBackedRowContainer` to not create the in-memory one and spill to disk right away, later I got rid off it but forgot to remove this change.
___
*[pkg/sql/rowcontainer/row_container.go, line 521 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_3LUm3pGJuGjoNLPL:-LXfFdp2A4tJboqxXm26:b4lleqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L521)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

As stated, the `DiskRowContainer` does this. Maybe the comment needs to be expanded?
</blockquote></details>

I expanded the comment, please let me know if additional info there would useful.
___
*[pkg/sql/rowcontainer/row_container.go, line 529 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZUQxD2hDFUjL6T4F_:-LX_L0v633SBgMgi6kNg:b-6roj95) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L529)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Do you need these two fields? Can you use the equivalent ones from within DiskBackedRowContainer?
</blockquote></details>

`DiskBackedRowContainer` doesn't have these fields explicitly. The underlying `MemRowContainer` has `datumAlloc` but no `rowAlloc`, and the underlying `DiskRowContainer` has neither of those. I think these fields need to be a part of `DiskBackedIndexedRowContainer` since it is this container's responsibility to cache the rows and decode datums as needed.
___
*[pkg/sql/rowcontainer/row_container.go, line 646 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZoV137UjCuqbH1yUO:-LXfFpO891BUtL5u3J2r:b-s8cou5) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L646)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

As I understand it the problem statement is this (focusing on disk): a row container must be created with an associated row ordinal and an ordering. The `DiskRowContainer` already does this with `rowID` and a supplied ordering. The difficulty however then comes from wanting to change the ordering on the `DiskRowContainer` but preserve the initial ordinal. I'm wondering if maybe we could rewrite keys in the DiskRowContainer, leaving the ordinal untouched (it is added at the end of the key). By rewrite I mean, when `ChangeOrdering` is called on the `DiskRowContainer`, we iterate over the `DiskRowContainer` and reinsert the row according to the given ordering. Don't know how possible it is.

Also, I think this could be a property observed on a normal `DiskBackedRowContainer`, maybe eliminating the need to create a new row container? Haven't thought much about how these guarantees would translate to memory row  containers.
</blockquote></details>

I implemented an approach with utilizing `rowID`s (the code can be seen here: https://github.com/yuzefovich/cockroach/commit/3e2cfaaef6c13e1cce847ec3a423cce7ed2c6308) , high level idea: when we spill to disk, we reuse `rowID`s that are assigned to each row when it is first added, so when we change ordering, we keep exactly the same ordinals (which are unique anyway, so there should be no problem with the loss of duplicated rows). When we’re in-memory, we need to keep a slice of indices separately and just need to make sure that indices are swapped accordingly when the rows are sorted.

As discussed with Alfonso, the initial approach seems better, so I decided to keep it.
___
*[pkg/sql/rowcontainer/row_container.go, line 679 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXZmwVs2J7zcLJ0G6dI:-LXfGQ05C1TZ3x-VEi9r:b-hnbxpl) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L679)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would have this function return an error rather than panic
</blockquote></details>

Good idea. I opened a PR that adds an option of returning an error to the interfaces, and I will do that here once that PR is merged.
___
*[pkg/sql/rowcontainer/row_container.go, line 683 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_53pw7SbJao83Y7R5:-LXdosmn8uKuidxw0xZc:b-vtqw96) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L683)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

How will this function be used? I wonder if it makes sense to just call `Next()` `n` times. We also have to be careful about the cache's memory overhead. 
</blockquote></details>

If you're curious what is the access pattern  for this `GetRow()` method will be, it is as follows: first, we need to compute the boundaries of the peer group, so the accesses will be for `prevPeerGroupEndIdx + 1,  prevPeerGroupEndIdx + 2, ... ` until we get to the row from the next peer group (or we run out of rows), then we compute the window function over the peer group and we will be accessing `GetRow()` again to get the index of rows at positions `prevPeerGroupEndIdx + 1,  prevPeerGroupEndIdx + 2, ... `. So, briefly, we iterate over the peer group twice and then move on to the next peer group. (As an aside: "peers" are rows that are equal on the ordering columns.)

Yes, I need to add memory accounting for the cache.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (changes requested):
> Reviewed 3 of 12 files at r3, 3 of 3 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/rowcontainer/disk_row_container.go, line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXtG2Wu-KIh5kNuiOpc:-LXtG2Wu-KIh5kNuiOpd:b3x4568) ([raw file](https://github.com/cockroachdb/cockroach/blob/741a790bbad7618786c91d3c88dc8183c36f5f7f/pkg/sql/rowcontainer/disk_row_container.go#L119)):*
> ```Go
> 		// pkg/sql/parser/datum.go.
> 		if _, ok := orderingIdxs[i]; !ok || sqlbase.HasCompositeKeyEncoding(d.types[i].SemanticType) || sqlbase.MustBeValueEncoded(d.types[i].SemanticType) {
> 			d.valueIdxs = append(d.valueIdxs, i)
> ```

Could you remind me what undesired behavior you were seeing here? It's a bit surprising that a datum type that `MustBeValueEncoded` was specified in the ordering indices, this might lead to incorrect assumptions on the part of the caller (i.e. I can order by `JSONB` columns)
___
*[pkg/sql/rowcontainer/row_container.go, line 683 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_53pw7SbJao83Y7R5:-LXtHOSI6BBIwv-MKJC0:bmobqcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L683)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

If you're curious what is the access pattern  for this `GetRow()` method will be, it is as follows: first, we need to compute the boundaries of the peer group, so the accesses will be for `prevPeerGroupEndIdx + 1,  prevPeerGroupEndIdx + 2, ... ` until we get to the row from the next peer group (or we run out of rows), then we compute the window function over the peer group and we will be accessing `GetRow()` again to get the index of rows at positions `prevPeerGroupEndIdx + 1,  prevPeerGroupEndIdx + 2, ... `. So, briefly, we iterate over the peer group twice and then move on to the next peer group. (As an aside: "peers" are rows that are equal on the ordering columns.)

Yes, I need to add memory accounting for the cache.
</blockquote></details>

I would run some microbenchmarks without the cache to see if it actually makes a difference, I would assume that caching behavior already exists lower down in the stack, might be interesting to try it out.
___
*[pkg/sql/rowcontainer/row_container.go, line 414 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXtGpnM6lKSttmi0pXN:-LXtGpnM6lKSttmi0pXO:bf3kpx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/741a790bbad7618786c91d3c88dc8183c36f5f7f/pkg/sql/rowcontainer/row_container.go#L414)):*
> ```Go
> 		return err
> 	}
> 	switch c := f.src.(type) {
> ```

Do you still need to do this switch?
___
*[pkg/sql/rowcontainer/row_container_test.go, line 379 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXtL3dbDQ9nPfZXYZ1C:-LXtL3dbDQ9nPfZXYZ1D:btylcge) ([raw file](https://github.com/cockroachdb/cockroach/blob/741a790bbad7618786c91d3c88dc8183c36f5f7f/pkg/sql/rowcontainer/row_container_test.go#L379)):*
> ```Go
> }
> 
> func TestDiskBackedIndexedRowContainer(t *testing.T) {
> ```

I think it would also be nice to see some `GetRow` tests.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/rowcontainer/disk_row_container.go, line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXtG2Wu-KIh5kNuiOpc:-LXtqXl94-kceISCocKW:bbwjw4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/741a790bbad7618786c91d3c88dc8183c36f5f7f/pkg/sql/rowcontainer/disk_row_container.go#L119)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Could you remind me what undesired behavior you were seeing here? It's a bit surprising that a datum type that `MustBeValueEncoded` was specified in the ordering indices, this might lead to incorrect assumptions on the part of the caller (i.e. I can order by `JSONB` columns)
</blockquote></details>

In our logic tests we have this query `SELECT avg(k) OVER (PARTITION BY kv.*) FROM kv ORDER BY 1` - a window function that uses a tuple for PARTITION BY clause. The behavior that I was seeing was as follows: I forced `hashRowContainer` to spill to disk, the tuple consisting of all columns for each row was used to encode into the key, and I believe it was being put in the correct order into the disk map. However, later, when I was trying to read the rows back from disk, I encountered an error with decoding. From what I understood, the tuples were encoded with very few bytes (like 1 byte each) instead of the whole 20 or so bytes for the whole tuple since it was sufficient to key encode it.

After discussing with Jordan, I'll make this change with the later window functions spilling PR since it utilizes this case, and it will be easier to test. We think that current implementation of `MustBeValueEncoded` is incorrect - tuples with key-encodable types can be encoded I believe, so I might open a different PR for that.
___
*[pkg/sql/rowcontainer/row_container.go, line 683 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_53pw7SbJao83Y7R5:-LXtsvPCCxS6MHHaoYqy:b-jryp6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L683)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I would run some microbenchmarks without the cache to see if it actually makes a difference, I would assume that caching behavior already exists lower down in the stack, might be interesting to try it out.
</blockquote></details>

It definitely made big difference on the trackernetwork query. When I ran that query without the cache, it spilled to disk and took about 6 minutes, so I started thinking about the access pattern and how to improve the performance. I saw the access pattern as described above and decided to add a cache for it. Once I implemented it (and tuned the cache size), the runtime decreased to 21 seconds.

I've just added a benchmark, and the results are:
```
BenchmarkDiskBackedIndexedRowContainer/InMemory-12             20000000           101 ns/op      78.71 MB/s
BenchmarkDiskBackedIndexedRowContainer/OnDiskWithCache-12       5000000           361 ns/op      22.12 MB/s
BenchmarkDiskBackedIndexedRowContainer/OnDiskWithoutCache-12               500000         39870 ns/op       0.20 MB/s
```
___
*[pkg/sql/rowcontainer/row_container.go, line 414 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXtGpnM6lKSttmi0pXN:-LXtsjQtAIH4crQXeAJD:brbedu) ([raw file](https://github.com/cockroachdb/cockroach/blob/741a790bbad7618786c91d3c88dc8183c36f5f7f/pkg/sql/rowcontainer/row_container.go#L414)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Do you still need to do this switch?
</blockquote></details>

You're right, I think we don't need it anymore since we're updating the container "in-place", and references to it don't need to be updated.
___
*[pkg/sql/rowcontainer/row_container_test.go, line 379 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXtL3dbDQ9nPfZXYZ1C:-LXttkfk25CAk0kLwDYK:b45459i) ([raw file](https://github.com/cockroachdb/cockroach/blob/741a790bbad7618786c91d3c88dc8183c36f5f7f/pkg/sql/rowcontainer/row_container_test.go#L379)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think it would also be nice to see some `GetRow` tests.
</blockquote></details>

`GetRow` is implicitly tested in `SpillingHalfway`.  I've just added an extra test for more explicit and full test of GetRow as well.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
> Reviewed 1 of 12 files at r3, 1 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/rowcontainer/disk_row_container.go, line 119 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXtG2Wu-KIh5kNuiOpc:-LXyqtGAAO8neizyQvUY:b-8qfbio) ([raw file](https://github.com/cockroachdb/cockroach/blob/741a790bbad7618786c91d3c88dc8183c36f5f7f/pkg/sql/rowcontainer/disk_row_container.go#L119)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

In our logic tests we have this query `SELECT avg(k) OVER (PARTITION BY kv.*) FROM kv ORDER BY 1` - a window function that uses a tuple for PARTITION BY clause. The behavior that I was seeing was as follows: I forced `hashRowContainer` to spill to disk, the tuple consisting of all columns for each row was used to encode into the key, and I believe it was being put in the correct order into the disk map. However, later, when I was trying to read the rows back from disk, I encountered an error with decoding. From what I understood, the tuples were encoded with very few bytes (like 1 byte each) instead of the whole 20 or so bytes for the whole tuple since it was sufficient to key encode it.

After discussing with Jordan, I'll make this change with the later window functions spilling PR since it utilizes this case, and it will be easier to test. We think that current implementation of `MustBeValueEncoded` is incorrect - tuples with key-encodable types can be encoded I believe, so I might open a different PR for that.
</blockquote></details>

Thanks for explaining. So essentially we were encoding a tuple as a key but not necessarily expecting an ordering. We will probably have to distinguish between these use cases when we revisit this.
___
*[pkg/sql/rowcontainer/row_container.go, line 683 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LX_53pw7SbJao83Y7R5:-LXyrD31AQEID6Ws2gpW:bcnb1oc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8d649f121c2a7b2679b66cfb4954c99e2c6a2f/pkg/sql/rowcontainer/row_container.go#L683)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It definitely made big difference on the trackernetwork query. When I ran that query without the cache, it spilled to disk and took about 6 minutes, so I started thinking about the access pattern and how to improve the performance. I saw the access pattern as described above and decided to add a cache for it. Once I implemented it (and tuned the cache size), the runtime decreased to 21 seconds.

I've just added a benchmark, and the results are:
```
BenchmarkDiskBackedIndexedRowContainer/InMemory-12             20000000           101 ns/op      78.71 MB/s
BenchmarkDiskBackedIndexedRowContainer/OnDiskWithCache-12       5000000           361 ns/op      22.12 MB/s
BenchmarkDiskBackedIndexedRowContainer/OnDiskWithoutCache-12               500000         39870 ns/op       0.20 MB/s
```
</blockquote></details>

Wow, this is definitely surprising to me. Thanks for running the benchmarks. This might be a motivation to add caching in the DiskRowContainer.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@jordanlewis PTAL when you get a chance.

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/rowcontainer/row_container.go, line 580 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXzJi7E-pXpvyf0OVW-:-LXzJi7E-pXpvyf0OVW0:bd3bfi0) ([raw file](https://github.com/cockroachdb/cockroach/blob/665a4cf3f26a7e1de6deec8bf8b6524245d49210/pkg/sql/rowcontainer/row_container.go#L580)):*
> ```Go
> // AddRow implements SortableRowContainer.
> func (f *DiskBackedIndexedRowContainer) AddRow(ctx context.Context, row sqlbase.EncDatumRow) error {
> 	// TODO(yuzefovich): I'm storing here the context which I need in GetRow().
> ```

I'm not sure what to do about this. @asubiotto @jordanlewis thoughts?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> Reviewed 1 of 3 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/rowcontainer/row_container.go, line 580 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXzJi7E-pXpvyf0OVW-:-LXzK_rCC9_xQyCODjoC:b-ys3kc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/665a4cf3f26a7e1de6deec8bf8b6524245d49210/pkg/sql/rowcontainer/row_container.go#L580)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I'm not sure what to do about this. @asubiotto @jordanlewis thoughts?
</blockquote></details>

GetRow should get a context, yes.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/rowcontainer/row_container.go, line 580 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LXzJi7E-pXpvyf0OVW-:-LXzzYtu6WrGcDpJEo1v:b-dtkdrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/665a4cf3f26a7e1de6deec8bf8b6524245d49210/pkg/sql/rowcontainer/row_container.go#L580)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

GetRow should get a context, yes.
</blockquote></details>

Done, thanks.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
Really nice work, and excellent tests. This is looking just about ready.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-:-LY0QIk29x88LJjP48FP:b5h0et5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/rowcontainer/row_container.go, line 701 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0O4FuBR2jwQQbTuGI:-LY0O4Fv1ReroSpr6CZ2:bf49nut) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container.go#L701)):*
> ```Go
> 						usage := sizeOfInt + int64(f.indexedRowsCache.GetFirst().(IndexedRow).Row.Size())
> 						f.indexedRowsCache.RemoveFirst()
> 						f.cacheMemAcc.Shrink(ctx, usage)
> ```

I'm surprised this row-by-row growing and shrinking didn't show up in the profile. It might be a bottleneck to do so, especially since we have the chance to do it a chunk at a time during this cache population step. Take a look at the profile and see if you see something here. Either way I'm also eager to get this phase of the work done, so if you do see something, note it down in an issue and we can move on anyway unless it's super egregious.
___
*[pkg/sql/rowcontainer/row_container.go, line 711 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0PH4_-EV_NsZ4RBAZ:-LY0PH4_-EV_NsZ4RBA_:bivu9t1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container.go#L711)):*
> ```Go
> 					}
> 					// We actually need to copy the row into memory.
> 					ir := IndexedRow{int(*rowIdx), f.rowAlloc.CopyRow(row)}
> ```

I see a fairly important performance optimization here. Your cache is a ring buffer. That means the memory in it is preserved somehow - you keep overwriting old cells. Doesn't this also mean that we have the chance to reuse the old allocated EncDatumRow memory that each old cell has a reference to, by the time we're going to overwrite one of them?

Seems like we should be allocating at most maxIndexedRowsCacheSize new EncDatumRows with the row alloc, and after that, reusing the old ones that we've allocated since they're all guaranteed to be the same size. I do think you'll probably see a pretty good perf improvement if you do this - my guess is that this CopyRow represents a very large proportion of the allocations in this component.

Happy to see this go in later as it might be a little tricky to get right.
___
*[pkg/sql/rowcontainer/row_container.go, line 728 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0ObCO1yz5BvBcBXqX:-LY0ObCO1yz5BvBcBXqY:bokllyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container.go#L728)):*
> ```Go
> 		return IndexedRow{int(*rowIdx), rowWithIdx[:len(rowWithIdx)-1]}, nil
> 	}
> 	return nil, errors.Errorf("unexpected last column: should be integer pos")
> ```

include details about what went wrong, like maybe the type of the thing you found instead of an int.
___
*[pkg/sql/rowcontainer/row_container_test.go, line 709 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0QBCnAAtKwgvX3vN7:-LY0QBCnAAtKwgvX3vN8:b28j2kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container_test.go#L709)):*
> ```Go
> 		}
> 	}
> 	return 0, nil
> ```

Kind of surprised this code doesn't already exist somewhere you can reuse, but looks good


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/rowcontainer/row_container.go, line 701 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0O4FuBR2jwQQbTuGI:-LY2Otwb9fKv8_jYz-e8:btktirx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container.go#L701)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm surprised this row-by-row growing and shrinking didn't show up in the profile. It might be a bottleneck to do so, especially since we have the chance to do it a chunk at a time during this cache population step. Take a look at the profile and see if you see something here. Either way I'm also eager to get this phase of the work done, so if you do see something, note it down in an issue and we can move on anyway unless it's super egregious.
</blockquote></details>

By "chunk at a time" do you mean to grow the account only when the underlying ring buffer increases in size?

I just did the profiling, neither `Grow` nor `Shrink` showed up (possibly, I did something wrong, but I don't see those neither through `web` nor through `focus=...`). Also, I believe `row.Size()` can change on a per row basis, so if we want to have somewhat exact memory accounting for the cache, we need to keep track of differences in sizes and occasionally update the memory account accordingly. One idea is to maintain the delta between what we actually are using at the moment and what we grew the memory account to and to update the memory account only when we exceed some threshold. But since this growing-shrinking pattern doesn't seem to give us a performance hit, maybe we should just keep it as is?
___
*[pkg/sql/rowcontainer/row_container.go, line 711 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0PH4_-EV_NsZ4RBAZ:-LY2SLF805Y_29UdLWcU:b9ikrgx) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container.go#L711)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I see a fairly important performance optimization here. Your cache is a ring buffer. That means the memory in it is preserved somehow - you keep overwriting old cells. Doesn't this also mean that we have the chance to reuse the old allocated EncDatumRow memory that each old cell has a reference to, by the time we're going to overwrite one of them?

Seems like we should be allocating at most maxIndexedRowsCacheSize new EncDatumRows with the row alloc, and after that, reusing the old ones that we've allocated since they're all guaranteed to be the same size. I do think you'll probably see a pretty good perf improvement if you do this - my guess is that this CopyRow represents a very large proportion of the allocations in this component.

Happy to see this go in later as it might be a little tricky to get right.
</blockquote></details>

Great point, thanks! When I wrote this code, I was briefly thinking about it, but I erroneously thought the row sizes could be different. We can (and definitely should) reuse that memory. This will require modifying `ring.Buffer`, and I will do so in a separate PR.
___
*[pkg/sql/rowcontainer/row_container.go, line 728 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0ObCO1yz5BvBcBXqX:-LY2Ndw44poVkD4bFhzj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container.go#L728)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

include details about what went wrong, like maybe the type of the thing you found instead of an int.
</blockquote></details>

Done.
___
*[pkg/sql/rowcontainer/row_container_test.go, line 709 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0QBCnAAtKwgvX3vN7:-LY2OFZM4PMr3dmrR0cY:b-cofqia) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container_test.go#L709)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Kind of surprised this code doesn't already exist somewhere you can reuse, but looks good
</blockquote></details>

Actually, I think this `sorter` is an exact copy of `partitionSorter` that lives in `windower.go`. I wonder whether I should export it from `rowcontainer` package and remove the one in `windower.go` (or the other way around)?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @yuzefovich)

___
*[pkg/sql/rowcontainer/row_container_test.go, line 709 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0QBCnAAtKwgvX3vN7:-LY2pQPz8c6o19Z_QIK-:bdenwpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container_test.go#L709)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Actually, I think this `sorter` is an exact copy of `partitionSorter` that lives in `windower.go`. I wonder whether I should export it from `rowcontainer` package and remove the one in `windower.go` (or the other way around)?
</blockquote></details>

Why don't you make a new package to put it and PeerGroupChecker and so on in? It's weird that that thing lives in sem/tree - make a package sql/window or something.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/rowcontainer/row_container_test.go, line 709 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-LY0QBCnAAtKwgvX3vN7:-LY3EWtg2fvLLTppgaI7:bxodp31) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8418e81022ad2c9b367d01cc7560bfd823c3595/pkg/sql/rowcontainer/row_container_test.go#L709)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Why don't you make a new package to put it and PeerGroupChecker and so on in? It's weird that that thing lives in sem/tree - make a package sql/window or something.
</blockquote></details>

Good point, I left a TODO. I attempted to create a new package, but it was pretty hairy, so I would like to merge this one and do the refactoring in a separate PR if you don't object.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the reviews! Some things talked about here are left as TODOs and will be addressed shortly in separate PRs.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/34196#-:-LY3jbMN2vb8JzvTy3uA:b-qvl8t9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)



<!-- Sent from Reviewable.io -->
